File: Lecture_Example/Week6/Bisect.hs
-- | A polynomial.
poly :: Double -> Double
poly x = (x*x*x) - x*x - 2


-- | Tail recusive bisection method.
bisect :: (Double -> Double) -> Double -> Double -> Double -> Double
bisect f tol a b =
    let c = (a + b) / 2
        y = f c
    in
        if abs y < tol
        then
            c
        else
            if signum (f a) == signum y
            then
                bisect f tol c b
            else
                bisect f tol a c


-- | Stream-based implementation of the bisection method.
bisections :: (Double -> Double) -> Double -> Double -> [Double]
bisections f a b =
    let c = (a + b) / 2
        y = f c
    in
        if signum (f a) == signum y
        then
            c : bisections f c b
        else
            c : bisections f a c


findSmall :: Double -> (Double -> Double) -> [Double] -> Double
findSmall tol f nums =
    head $ filter (\x -> abs (f x) < tol) nums

File: Lecture_Example/Week6/bisect.py
def bisect(f, tol, a, b):
  # Precondition: f(a) and f(b) have different signs.
  c = (a + b)/2
  while abs(f(c)) >= tol:
    if sign(f(a)) == sign(f(c)):
      a = c
    else:
      b = c
    c = (a + b)/2

  return c


def sign(n):
  if n > 0:
    return 1
  elif n == 0:
    return 0
  else:
    return -1
